.github/scripts/sketch_utils.sh

```diff
@@ -1,43 +1,155 @@
+    # Options default values
+
+    local fm_opt="qio"
+    local ff_opt="80"
+    local fs_opt="4M"
+    local partition_opt="huge_app"
+
+    local options=0
+
+    while [ ! -z "$1" ]; do
+        case "$1" in
+        -ai )
+            shift
+            ide_path=$1
+            ;;
+        -au )
+            shift
+            user_path=$1
+            ;;
+        -t )
+            shift
+            target=$1
+            ;;
+        -fqbn )
+            shift
+            fqbn=$1
+            ;;
+        -s )
+            shift
+            sketchdir=$1
+            ;;
+        -ff )
+            shift
+            ff_opt=$1
+            options=1
+            ;;
+        -fm )
+            shift
+            fm_opt=$1
+            options=1
+            ;;
+        -fs )
+            shift
+            fs_opt=$1
+            options=1
+            ;;
+        * )
+            break
+            ;;
+        esac
+        shift
+    done
+
+    xtra_opts=$*
+
+    if [ -z $sketchdir ]; then
+        echo "ERROR: Sketch directory not provided"
+        echo "$USAGE"
+        exit 1
     fi
 
-    local ide_path=$1
-    local usr_path=$2
-    local fqbn=$3
-    local sketch=$4
-    local xtra_opts=$5
-    local win_opts=$6
+    # No FQBN was passed, try to get it from other options
+
+    if [ -z $fqbn ]; then
+        if [ -z $target ]; then
+            echo "ERROR: Unspecified chip"
+            echo "$USAGE"
+            exit 1
+        fi
+
+        # The options are either stored in the test directory, for a per test
+        # customization or passed as parameters.  Command line options take
+        # precedence.  Note that the following logic also falls to the default
+        # parameters if no arguments were passed and no file was found.
+
+        if [ $options -eq 0 ] && [ -f $sketchdir/cfg.json ]; then
+            # The config file could contain multiple FQBNs for one chip.  If
+            # that's the case we build one time for every FQBN.
+
+            len=`jq -r --arg chip $target '.targets[] | select(.name==$chip) | .fqbn | length' $sketchdir/cfg.json`
+            fqbn=`jq -r --arg chip $target '.targets[] | select(.name==$chip) | .fqbn' $sketchdir/cfg.json`
+        else
+            # Since we are passing options, we will end up with only one FQBN to
+            # build.
+
+            len=1
+
+            # Select the common part of the FQBN based on the target.  The rest will be
+            # appended depending on the passed options.
+
+            case "$target" in
+                "esp32") fqbn="espressif:esp32:esp32:"
+                ;;
+                "esp32s2") fqbn="espressif:esp32:esp32s2:"
+                ;;
+                "esp32c3") fqbn="espressif:esp32:esp32c3:"
+                ;;
+                "esp32s3") fqbn="espressif:esp32:esp32s3:"
+                ;;
+            esac
+
+            partition="PartitionScheme=$partition_opt"
```

Comment:
I will suggest to extract the option names also from some argument. We should not be limited by these options, nor we should need to change them if we adjust the options in Arduino's configs. We can have the default FQBNs defined somewhere, if special is needed, then JSON should be provided.

Reply:
I only added the basics that (I thought) will be present in all chips.  For further customization one can pass a whole FQBN, this is not very different from passing options with their names.  It can save some typing if we extended options to the common parts, for instance for S3 the FQBN will start with `espressif:esp32:esp32s3:` and then we extend it with `./tests_build.sh -t esp32 -a "FlashMode=qio120,PartitionScheme=huge_app`, the resulting FQBN will be `espressif:esp32:esp32s3:FlashMode=qio120,PartitionScheme=huge_app`

Reply:
In this case those flash options are useless and cumbersome, I can just remove them.

Reply:
> In this case those flash options are useless and cumbersome, I can just remove them.

I did so.